Framework/Parallel/src/IO/EventParser.cpp

```diff
@@ -0,0 +1,198 @@
+void EventParser<IndexType, TimeZeroType, TimeOffsetType>::setPulseInformation(
+    const std::vector<IndexType> &event_index,
+    const std::vector<TimeZeroType> &event_time_zero) const {
+  m_eventIndex = event_index;
+  m_eventTimeZero = event_time_zero;
+}
+
+/** Used the detectorIDs supplied to calculate the corresponding global spectrum
+ * numbers using the bankOffsets stored at object creation.
+ * @param event_id_start Starting position of chunk of data containing detector
+ * IDs
+ * @param count Number of items in data chunk
+ * @param bankIndex Index into the list of bank offsets.
+ */
+template <class IndexType, class TimeZeroType, class TimeOffsetType>
+void EventParser<IndexType, TimeZeroType, TimeOffsetType>::
+    eventIdToGlobalSpectrumIndex(int32_t *event_id_start, size_t count,
+                                 size_t bankIndex) const {
+  m_globalSpectrumIndex.resize(count);
+
+  for (size_t i = 0; i < count; i++)
+    m_globalSpectrumIndex[i] = event_id_start[i] - m_bankOffsets[bankIndex];
+}
+
+/** Extracts event information from the list of time offsets and global spectrum
+ * indices using the event_index and event_time_offset tables provided from
+ * file. These events are separated according to MPI ranks.
+ * @param globalSpectrumIndex list of spectrum indices corresponding to tof data
+ * @param eventTimeOffset tof data
+ * @param offset File offset (index) for tof data. Used to track event_index and
+ * event_time_zero positions
+ */
+template <class IndexType, class TimeZeroType, class TimeOffsetType>
+void EventParser<IndexType, TimeZeroType, TimeOffsetType>::
+    extractEventsForRanks(const std::vector<int32_t> &globalSpectrumIndex,
+                          const TimeOffsetType *eventTimeOffset,
+                          size_t offset) {
+  // TODO: Remove when doing MPI
+  m_allRankData.resize(1);
+
+  for (auto &item : m_allRankData)
+    item.clear();
+
+  const auto count = globalSpectrumIndex.size();
+  // Determine start and end pulse/s for chunk of data provided.
+  auto result = findStartAndEndPulses<IndexType, TimeOffsetType>(
+      m_eventIndex, offset, count, m_posInEventIndex);
+
+  for (size_t pulse = result.first; pulse < result.second; ++pulse) {
+    const auto start =
+        std::max(offset, static_cast<hsize_t>(m_eventIndex[pulse])) - offset;
+    const auto end =
+        std::min(offset + count,
+                 static_cast<hsize_t>(pulse != m_eventIndex.size() - 1
+                                          ? m_eventIndex[pulse + 1]
+                                          : offset + count)) -
+        offset;
+
+    for (size_t i = start; i < end; ++i) {
+      // TODO: select appropriate ranks round robin?
+      // int rank = global_spectrum_index[i] % nrank;
+      // rank_data[rank].push_back(...)
+      int rank = 0;
+      m_allRankData[rank].push_back(
+          Event{globalSpectrumIndex[i],
+                TofEvent{static_cast<double>(eventTimeOffset[i]),
```

Comment:
I think I would do the translation into `TofEvent` on the receiver side. `TofEvent` has 16 bytes (aligned to 8 byte), so that brings `Event` to 24 bytes. If you keep the original types it might be only 16 byte (if `event_time_offset` is 4 byte). Also it may be the case that we use not all ranks for loading, so as much work as possible should be done on the receiver side.